### ISSUE TITLE
MSSQL Integration: transaction_log metrics should load from all tables

### ISSUE THREAD

**Issue by lalit-satapathy** [elastic/integrations#4108]
The MSSQL Integration currently loads "transaction_log metrics" from following tables:
- tempdb
- msdb
- model
- master

The integration code should be updated to load from all the tables.

**Comment by lalit-satapathy** [comment]
The short-term solution to this would be to add an optional flag in transaction_log for user to enter a customer DB name. The metrics can be extracted for this customer DB along with all system DBs. CC @ManojS-shetty

**Comment by VF-JustinMolnar** [comment]
Hi @lalit-satapathy ,
I am interested in the short term solution you describe. Do you have any details on how we can add an optional input to the transaction_log so we can capture our custom databases?
Thanks.

**Comment by ManojS-shetty** [comment]
> Hi @lalit-satapathy , I am interested in the short term solution you describe. Do you have any details on how we can add an optional input to the transaction_log so we can capture our custom databases? Thanks.

HI @VF-JustinMolnar We would like to inform you that as Lalit suggested we will be providing this optional input in transaction_log  as a part of package release. 

We will be updating the progress of the development and testing on this issue thread.

Thank you

**Comment by Danouchka** [comment]
Hi @lalit-satapathy  for me it's a bug compared to Metricbeat capabilities , Customer called it a regression. Why on that point, Metricbeat retrieves metrics for all DBs and Elastic-Agent only system DBs ? 
Additional input , even on short term, does not make sense because it supposes to enter manually each DB name. 
But the idea is also to monitor Customer DBs for which the DBA or infra team would not be aware of.

**Comment by lalit-satapathy** [comment]
Hi @Danouchka,

This is only a "short term solution" as mentioned in [here](https://github.com/elastic/integrations/issues/4108#issuecomment-1239395296). The long-term solution is TBD.

**Comment by ritalwar** [comment]
Hi, similar to metricbeat, I am working on retrieving metrics for all the DBs using agent as well. 
Will be updating the progress here.

**Comment by ritalwar** [comment]
Short-term solution where user database can be manually provided is done. https://github.com/elastic/integrations/pull/4869.

**Comment by rameshelastic** [comment]
@ritalwar , lets update the latest on this issue.

**Comment by ritalwar** [comment]
Working on long term solution approach where we can get all user DBs metrics for transaction log.
Able to resolve it for sys.dm_db_log_stats table data and able to get data for all the DBs.
Trying to find a way to get all Dbs data from sys.dm_db_log_space_usage table.

**Comment by ritalwar** [comment]
Considering a strategy where a first integration can be added to obtain a list of all databases. The user can then edit the integration to add database names from that list and obtain the corresponding metrics.

**Comment by Danouchka** [comment]
Hi 

In that case, when you retrieve the list of all database, you have also to get some useful information (asked by L’Oreal DBA) such as if tthey are online or not, their file path, the occupied size of the file etc..
I had written an ER for that here https://github.com/elastic/integrations/issues/4260 

But why not retrieving all of databases metrics ? 
For instance, When you retrieve metrics from GCE Compute and the customer has  thousands of VM,  we dont enter the hostnames we want to monitor one by one in the GCP integration 
So why would it be different approach here ? 

Thanks, Dan 


Dan Abitbol
Solutions Architect SEMEA, elastic.co
Search. Observe. Protect.
+33662168617

￼

**Comment by Danouchka** [comment]
Hi  

Your strategy may be good but please 

1/ explain why it's different from GCP compute metrics 
2/ do you agree then that we have to implement [#4260](https://github.com/elastic/integrations/issues/4260) before 

Thank you

**Comment by ritalwar** [comment]
Hi,

I'm not certain about the comparison, so currently looking into the other possibilities as well.

Additionally, as mentioned earlier, trying to find a means to collect data for all Dbs from the sys.dm_db_log_space_usage table, making some headway with the use of the "DBCC SQLPERF (LOGSPACE)" query, which is giving Log Size (MB) and Log Space Used (%) data for all the dbs.

Trying to further get  total_log_size_in_bytes and log_space_in_bytes_since_last_backup data for all databases.

**Comment by Danouchka** [comment]
Hi @ritalwar 

Please combine your query with the latest one mentionned in https://github.com/elastic/integrations/issues/4260 
As we need those information (path to filename, online status) as well. Thank you 

For reminder, The query mentionned in https://github.com/elastic/integrations/issues/4260. (just one issue with this one below, SpaceUsed is null for user DBs) 
`SELECT DB_NAME() AS DbName,
        name AS FileName,
       type_desc,
        size/128.0 AS CurrentSizeMB, 
        size/128.0 - CAST(FILEPROPERTY(name, ''SpaceUsed'') AS INT)/128.0 AS FreeSpaceMB,
        State_desc as State_desc,
        physical_name as physical_name,
        (size/128.0 - CAST(FILEPROPERTY(name, ''SpaceUsed'') AS INT)/128.0)*100./(size/128.0 ) as percentfree
FROM sys.master_files WHERE type IN (0,1); `

**Comment by Danouchka** [comment]
Hello

Regarding latest elastic-agent integration "**Microsoft SQL Server**" , version 1.13.0, we have still issues.
Although we can add user databases, some metrics are not retrieved well regarding the transaction log

In elastic-integration, I added the 2 following user DBs 
- Microsoft.eShopOnWeb.Identity
- Microsoft.eShopOnWeb.CatalogDb 

I have 2 issues regarding transaction log metrics 
a) I have part of the metrics for Microsoft.eShopOnWeb.Identity. You can compare with metricbeat.   
_Elastic agent integration_ 
<img width="1965" alt="Capture d’écran 2023-02-27 à 12 46 56" src="https://user-images.githubusercontent.com/48563314/221555902-55f78017-83d5-4d61-83ef-692e82314138.png">
_Metricbeat mssql module_ 
<img width="1983" alt="Capture d’écran 2023-02-27 à 12 47 04" src="https://user-images.githubusercontent.com/48563314/221555937-72908227-a189-4fd7-afe3-510915a85956.png">

b) I have data only for Microsoft.eShopOnWeb.Identity and not Microsoft.eShopOnWeb.CatalogDb  
_Elastic agent integration_ 
<img width="1959" alt="Capture d’écran 2023-02-27 à 12 49 06" src="https://user-images.githubusercontent.com/48563314/221556340-8faaa614-e1a5-4aca-b184-2b8967ff115a.png">
_Metricbeat mssql module_ 
<img width="1947" alt="Capture d’écran 2023-02-27 à 12 49 13" src="https://user-images.githubusercontent.com/48563314/221556397-c6e1daaf-4525-449d-b481-f7131528389d.png">

c) In Elastic Agent integration, I have the following error 
> [elastic_agent.metricbeat][error] Error fetching data for metricset sql.query: fetch table mode failed: mssql: Database ‘Microsoft’ does not exist. Make sure that the name is entered correctly.

Does elastic agent have issues with User DB dotted names ? 

Elastic Agent diagnostics available at 
`curl -L -H 'Authorization: 8944d932a228f794' -o 'elastic-agent-diagnostics-2023-02-27T11-51-45Z-00.zip' https://upload.elastic.co/d/82e499b0fc5d2b72464f3ce5ec77034be3c1c28978641255a971d525c5f5919a `

**Comment by Danouchka** [comment]
My integration settings
cc @lalit-satapathy 
<img width="953" alt="Capture d’écran 2023-02-27 à 14 20 09" src="https://user-images.githubusercontent.com/48563314/221576914-ca4dbd37-5957-434b-97bd-8f8a7b806d5f.png">

**Comment by ritalwar** [comment]
Separate issue filed to support user dbs with special characters.
Please track here: https://github.com/elastic/integrations/issues/5528

**Comment by rameshelastic** [comment]
@ritalwar , can we close this too?

**Comment by ritalwar** [comment]
We currently support user databases in such a way that customers can add a list of user databases to observe performance and transaction-log metrics.
However, it would take some time to support all user databases by default;   @lalit-satapathy can we create a separate issue to provide that support, or should we keep this issue open and decide priority for this in 8.9?

**Comment by shmsr** [comment]
> Hi @ritalwar
> 
> Please combine your query with the latest one mentionned in #4260 As we need those information (path to filename, online status) as well. Thank you
> 
> For reminder, The query mentionned in #4260. (just one issue with this one below, SpaceUsed is null for user DBs) `SELECT DB_NAME() AS DbName, name AS FileName, type_desc, size/128.0 AS CurrentSizeMB, size/128.0 - CAST(FILEPROPERTY(name, ''SpaceUsed'') AS INT)/128.0 AS FreeSpaceMB, State_desc as State_desc, physical_name as physical_name, (size/128.0 - CAST(FILEPROPERTY(name, ''SpaceUsed'') AS INT)/128.0)*100./(size/128.0 ) as percentfree FROM sys.master_files WHERE type IN (0,1); `

For user database(s), querying from `sys.master_files` and getting `NULL` for the concerned filled seems expected. From the documentation of `sys.master_files`:

> Contains a row per file of a database as stored in the master database.

So it does not say user databases and that's why I think you are getting `NULL`.  To get the relevant info I think you have to use `sys.database_files` instead.

Example: 

- https://learn.microsoft.com/en-us/azure/azure-sql/database/file-space-manage?view=azuresql#database-data-space-allocated-and-unused-allocated-space
- https://learn.microsoft.com/en-us/sql/relational-databases/system-catalog-views/sys-database-files-transact-sql?view=sql-server-ver16#examples

cc: @Danouchka

**Comment by Danouchka** [comment]
Hi @shmsr I will give a try thank you ! and let you know

**Comment by Danouchka** [comment]
I get this below. 
Strange, I get some NULL regardless using sys.database_files or sys.master_files. 


 ```
sqlcmd -S localhost -U SA -P sqlMS200675# -Q "SELECT DB_NAME() AS DbName, name AS FileName, type_desc, size/128.0 AS CurrentSizeMB, size/128.0 - CAST(FILEPROPERTY(name, 'SpaceUsed') AS bigint)/128.0 AS FreeSpaceMB, State_desc as State_desc, physical_name as physical_name, CAST(max_size AS bigint)/128.0 AS max_file_size_mb, (size/128.0 - CAST(FILEPROPERTY(name, 'SpaceUsed') AS bigint)/128.0)*100./(size/128.0 ) as percentfree FROM sys.master_files  WHERE type_desc IN ('ROWS','LOG') ;"
DbName                                                                                                                           FileName                                                                                                                         type_desc                                                    CurrentSizeMB       FreeSpaceMB                   State_desc                                                   physical_name                                                                                                                                                                                                                                                        max_file_size_mb             percentfree
-------------------------------------------------------------------------------------------------------------------------------- -------------------------------------------------------------------------------------------------------------------------------- ------------------------------------------------------------ ------------------- ----------------------------- ------------------------------------------------------------ -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ---------------------------- ----------------------------------------
master                                                                                                                           master                                                                                                                           ROWS                                                                    4.437500                       .812500 ONLINE                                                       /var/opt/mssql/data/master.mdf                                                                                                                                                                                                                                                           -.007812                               18.3098591
master                                                                                                                           mastlog                                                                                                                          LOG                                                                     2.000000                      1.351563 ONLINE                                                       /var/opt/mssql/data/mastlog.ldf                                                                                                                                                                                                                                                          -.007812                               67.5781500
master                                                                                                                           tempdev                                                                                                                          ROWS                                                                    8.000000                          NULL ONLINE                                                       /var/opt/mssql/data/tempdb.mdf                                                                                                                                                                                                                                                           -.007812                                     NULL
master                                                                                                                           templog                                                                                                                          LOG                                                                     8.000000                          NULL ONLINE                                                       /var/opt/mssql/data/templog.ldf                                                                                                                                                                                                                                                          -.007812                                     NULL
master                                                                                                                           tempdev2                                                                                                                         ROWS                                                                    8.000000                          NULL ONLINE                                                       /var/opt/mssql/data/tempdb2.ndf                                                                                                                                                                                                                                                          -.007812                                     NULL
master                                                                                                                           tempdev3                                                                                                                         ROWS                                                                    8.000000                          NULL ONLINE                                                       /var/opt/mssql/data/tempdb3.ndf                                                                                                                                                                                                                                                          -.007812                                     NULL
master                                                                                                                           tempdev4                                                                                                                         ROWS                                                                    8.000000                          NULL ONLINE                                                       /var/opt/mssql/data/tempdb4.ndf                                                                                                                                                                                                                                                          -.007812                                     NULL
master                                                                                                                           modeldev                                                                                                                         ROWS                                                                    8.000000                          NULL ONLINE                                                       /var/opt/mssql/data/model.mdf                                                                                                                                                                                                                                                            -.007812                                     NULL
master                                                                                                                           modellog                                                                                                                         LOG                                                                     8.000000                          NULL ONLINE                                                       /var/opt/mssql/data/modellog.ldf                                                                                                                                                                                                                                                         -.007812                                     NULL
master                                                                                                                           MSDBData                                                                                                                         ROWS                                                                   13.437500                          NULL ONLINE                                                       /var/opt/mssql/data/MSDBData.mdf                                                                                                                                                                                                                                                         -.007812                                     NULL
master                                                                                                                           MSDBLog                                                                                                                          LOG                                                                      .500000                          NULL ONLINE                                                       /var/opt/mssql/data/MSDBLog.ldf                                                                                                                                                                                                                                                    2097152.000000                                     NULL
master                                                                                                                           Microsoft.eShopOnWeb.CatalogDb                                                                                                   ROWS                                                                    8.000000                          NULL ONLINE                                                       /var/opt/mssql/data/Microsoft.eShopOnWeb.CatalogDb.mdf                                                                                                                                                                                                                                   -.007812                                     NULL
master                                                                                                                           Microsoft.eShopOnWeb.CatalogDb_log                                                                                               LOG                                                                     8.000000                          NULL ONLINE                                                       /var/opt/mssql/data/Microsoft.eShopOnWeb.CatalogDb_log.ldf                                                                                                                                                                                                                         2097152.000000                                     NULL
master                                                                                                                           Microsoft.eShopOnWeb.Identity                                                                                                    ROWS                                                                    8.000000                          NULL ONLINE                                                       /var/opt/mssql/data/Microsoft.eShopOnWeb.Identity.mdf                                                                                                                                                                                                                                    -.007812                                     NULL
master                                                                                                                           Microsoft.eShopOnWeb.Identity_log                                                                                                LOG                                                                     8.000000                          NULL ONLINE                                                       /var/opt/mssql/data/Microsoft.eShopOnWeb.Identity_log.ldf                                                                                                                                                                                                                          2097152.000000                                     NULL

(15 rows affected)
[root@sa-da-webserver ~]# sqlcmd -S localhost -U SA -P sqlMS200675# -Q "SELECT DB_NAME() AS DbName, name AS FileName, type_desc, size/128.0 AS CurrentSizeMB, size/128.0 - CAST(FILEPROPERTY(name, 'SpaceUsed') AS bigint)/128.0 AS FreeSpaceMB, State_desc as State_desc, physical_name as physical_name, CAST(max_size AS bigint)/128.0 AS max_file_size_mb, (size/128.0 - CAST(FILEPROPERTY(name, 'SpaceUsed') AS bigint)/128.0)*100./(size/128.0 ) as percentfree FROM sys.database_files  WHERE type_desc IN ('ROWS','LOG') ;"
DbName                                                                                                                           FileName                                                                                                                         type_desc                                                    CurrentSizeMB       FreeSpaceMB                   State_desc                                                   physical_name                                                                                                                                                                                                                                                        max_file_size_mb             percentfree
-------------------------------------------------------------------------------------------------------------------------------- -------------------------------------------------------------------------------------------------------------------------------- ------------------------------------------------------------ ------------------- ----------------------------- ------------------------------------------------------------ -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ---------------------------- ----------------------------------------
master                                                                                                                           master                                                                                                                           ROWS                                                                    4.437500                       .812500 ONLINE                                                       /var/opt/mssql/data/master.mdf                                                                                                                                                                                                                                                           -.007812                               18.3098591
master                                                                                                                           mastlog                                                                                                                          LOG                                                                     2.000000                      1.351563 ONLINE                                                       /var/opt/mssql/data/mastlog.ldf                                                                                                                                                                                                                                                          -.007812                               67.5781500
```

(2 rows affected)
[root@sa-da-webserver ~]#

**Comment by shmsr** [comment]
@Danouchka Do you have any user-defined DB? I just see master. Also, with the second command, there are just 2 rows and that too of master and I do not any `NULL` there. Do you mean something else?

**Comment by Danouchka** [comment]
Good Point ! 
Microsoft.eShopOnWeb.CatalogDb    was supposed to be a user DB ...let me check how it was created !

**Comment by shmsr** [comment]
To summarize, we already have a [solution](https://github.com/elastic/integrations/issues/4108#issuecomment-1239395296) in place which enables users to manually feed the user databases along with the already set system databases (as default). But to close this issue we’d ideally want to have a solution that gets all accessible databases on a server.

I have a proposal that suggests a simple approach to how we can achieve the same instead of introducing SQL chaining similar to that of HHTTP JSON chaining in beats. Implementation is quite simple and it'd help us close this issue with a solid and simple solution without affecting older functionality.

Also, even if introduce SQL chaining which will introduce some kind of chaining where we use the response of one query into another. Although, we have not decided on any low-level details on how that’s going to look so far. But in my opinion, it does not have many use cases. Considering, integrations only fetch metrics, logs, etc. (for instrumentation/ observability needs) from the databases; the requirement of some kind of chaining seems unnecessary. If someone has a solid use case, I am all ears.

First approach:

- `EXEC sp_MSforeachdb` and with command injection (`?`), give the database name to the command.
```sql
EXEC sp_MSforeachdb @command;
GO
```

Example:

```sql
/* Print all table names for each database. */
EXEC sp_MSforeachdb 'USE [?]; SELECT name from sys.tables';
GO
```

```sql
/* Print all db names: system and user dbs */
EXEC sp_MSforeachdb 'PRINT ''?''';
GO
```

Related: [Run the same command on all SQL server's databases without cursors](https://www.mssqltips.com/sqlservertip/1414/run-same-command-on-all-sql-server-databases-without-cursors/)

But researching more on `sp_MSforeachdb` I came across [this](https://eitanblumin.com/2021/08/05/simplest-alternative-to-sp_msforeachdb/) which recommends against using it and suggests an alternative which is not a good alternative for us.

So, here's my suggested approach:

- We will introduce a new field called "**Fetch from all databases**" in the concerned integration which is of type `bool` so we'd see that field as a **toggle switch** on the dashboard. When we toggle it on i.e., true then it'll override the databases set manually from the dashboard. If the toggle switch is off, the integration would still work as before i.e., fetch the metrics from manually set databases.
- In the `stream.yml.hbs` file, we'll use `if` and `else` helpers for the handlebars script so that we can do a conditional check on the field `fetch_from_all_dbs`. It would look similar to the following patch:

```patch
diff --git a/packages/microsoft_sqlserver/data_stream/transaction_log/agent/stream/stream.yml.hbs b/packages/microsoft_sqlserver/data_stream/transaction_log/agent/stream/stream.yml.hbs
index 71c6c2aa0..6aa16f27c 100644
--- a/packages/microsoft_sqlserver/data_stream/transaction_log/agent/stream/stream.yml.hbs
+++ b/packages/microsoft_sqlserver/data_stream/transaction_log/agent/stream/stream.yml.hbs
@@ -7,13 +7,22 @@ driver: mssql
 raw_data.enabled: true
 # Collect the transaction logs from the system database
 sql_queries:
-{{#if databases}}
-{{#each databases as |database_name i|}}
-  - query: "SELECT @@servername AS server_name, @@servicename AS instance_name, name As 'database_name', database_id FROM sys.databases WHERE name='{{database_name}}';"
-    response_format: table
-  - query: "SELECT @@servername AS server_name, @@servicename AS instance_name, name As 'database_name', l.database_id, l.total_log_size_mb, l.active_log_size_mb,l.log_backup_time,l.log_since_last_log_backup_mb,l.log_since_last_checkpoint_mb,l.log_recovery_size_mb from sys.dm_db_log_stats(DB_ID('{{database_name}}')) l INNER JOIN sys.databases s ON l.database_id = s.database_id WHERE s.database_id = DB_ID('{{database_name}}') ;"
-    response_format: table
-  - query: "USE [{{database_name}}] ; SELECT @@servername AS server_name, @@servicename AS instance_name, name As 'database_name', l.database_id, l.total_log_size_in_bytes As total_log_size_bytes, l.used_log_space_in_bytes As used_log_space_bytes, l.used_log_space_in_percent As used_log_space_pct, l.log_space_in_bytes_since_last_backup from sys.dm_db_log_space_usage l INNER JOIN sys.databases s ON l.database_id = s.database_id WHERE s.database_id = DB_ID('{{database_name}}') ;"
-    response_format: table
-{{/each}}
-{{/if}}
\ No newline at end of file
+{{#if fetch_from_all_dbs}}
+- query: "SELECT @@servername AS server_name, @@servicename AS instance_name, name As 'database_name', database_id FROM sys.databases WHERE name='{{database_name}}';"
+  response_format: table
+- query: "SELECT @@servername AS server_name, @@servicename AS instance_name, name As 'database_name', l.database_id, l.total_log_size_mb, l.active_log_size_mb,l.log_backup_time,l.log_since_last_log_backup_mb,l.log_since_last_checkpoint_mb,l.log_recovery_size_mb from sys.dm_db_log_stats(DB_ID('{{database_name}}')) l INNER JOIN sys.databases s ON l.database_id = s.database_id WHERE s.database_id = DB_ID('{{database_name}}') ;"
+  response_format: table
+- query: "USE [{{database_name}}] ; SELECT @@servername AS server_name, @@servicename AS instance_name, name As 'database_name', l.database_id, l.total_log_size_in_bytes As total_log_size_bytes, l.used_log_space_in_bytes As used_log_space_bytes, l.used_log_space_in_percent As used_log_space_pct, l.log_space_in_bytes_since_last_backup from sys.dm_db_log_space_usage l INNER JOIN sys.databases s ON l.database_id = s.database_id WHERE s.database_id = DB_ID('{{database_name}}') ;"
+  response_format: table
+{{#else}}
+  {{#if databases}}
+    {{#each databases as |database_name i|}}
+      - query: "SELECT @@servername AS server_name, @@servicename AS instance_name, name As 'database_name', database_id FROM sys.databases WHERE name='{{database_name}}';"
+        response_format: table
+      - query: "SELECT @@servername AS server_name, @@servicename AS instance_name, name As 'database_name', l.database_id, l.total_log_size_mb, l.active_log_size_mb,l.log_backup_time,l.log_since_last_log_backup_mb,l.log_since_last_checkpoint_mb,l.log_recovery_size_mb from sys.dm_db_log_stats(DB_ID('{{database_name}}')) l INNER JOIN sys.databases s ON l.database_id = s.database_id WHERE s.database_id = DB_ID('{{database_name}}') ;"
+        response_format: table
+      - query: "USE [{{database_name}}] ; SELECT @@servername AS server_name, @@servicename AS instance_name, name As 'database_name', l.database_id, l.total_log_size_in_bytes As total_log_size_bytes, l.used_log_space_in_bytes As used_log_space_bytes, l.used_log_space_in_percent As used_log_space_pct, l.log_space_in_bytes_since_last_backup from sys.dm_db_log_space_usage l INNER JOIN sys.databases s ON l.database_id = s.database_id WHERE s.database_id = DB_ID('{{database_name}}') ;"
+        response_format: table
+    {{/each}}
+  {{/if}}
+{{/if}}
```

Above, I have added some extra blocks as needed and tweaked one query out of the three queries i.e., I omitted `USE [{{database_name}}]`. Also, we are not iterating over databases anymore if `fetch_from_all_dbs` is set to `true`.

- We'll add a new field to the SQL module of the beats repository so that it also takes into consideration the new field introduced in the module in the `integration` repository. The new field would be called `FetchAllDB` of type `bool`. We'll add the field to this [struct](https://github.com/elastic/beats/blob/7eb4a224cc99c2f71e7c4c7d44c7e056dc094143/x-pack/metricbeat/module/sql/query/query.go#L50).
- Now we can fetch all databases names in each call to [Fetch](https://github.com/elastic/beats/blob/7eb4a224cc99c2f71e7c4c7d44c7e056dc094143/x-pack/metricbeat/module/sql/query/query.go#L112) so that we only query databases that are there in that point of time because it might happen that some databases were there during the previous scrape but they are not currently there during the current scrape.
- If `FetchAllDB` is `true`, then range over the databases and execute the respective queries. But every query must be prefixed with `USE <db_name>;` where `<db_name>` is the name of the databases that we get when we range over a set of databases.
```patch
- @command
+ USE <db_name>; @command
```
- Every query would work as expected and this approach would require minimal changes on both `integrations` and `beats` repository.

---

How to get the all databases?
 
There are a few solutions but they might have some differences in performance and other dynamics. We can pick what works best for us after doing some testing and benchmarking.

```sql
SELECT [name] FROM sys.databases;		/* simplest */
```

```sql
SELECT [name]
FROM sys.databases WITH (NOLOCK)
WHERE state = 0 				/* online only */
AND HAS_DBACCESS([name]) = 1; 			/* accessible only  */
```

```sql
SELECT [name]
FROM sys.databases WITH (NOLOCK)
WHERE state = 0 				/* online only */
AND HAS_DBACCESS([name]) = 1 			/* accessible only  */
AND database_id > 4 AND is_distributor = 0; 	/* ignore system databases */
```

The 2nd and 3rd commands are more appropriate because we want to only fetch databases that are accessible. 3rd command ignores the system databases, so ideally, we should go for 2nd. `NOLOCK` has its own pros and cons but for us pros do outweigh the cons:

> - The NOLOCK hint allows SQL to read data from tables by ignoring any locks and therefore not get blocked by other processes.
> - This can improve query performance by removing the blocks but introduces the possibility of dirty reads.

---

I'd like to know your feedback on the same.

cc: @lalit-satapathy @ritalwar

**Comment by shmsr** [comment]
I've updated the patch with the latest SQK queries we will use with the proposal given above. Please refer to the diff patch for the updated queries.

**Comment by shmsr** [comment]
The feature requested is released in the 2.2.0 version of Microsoft SQL Server. For more info see: https://github.com/elastic/integrations/pull/7500

Please ensure to use the 8.10 stack or above in order to use the `fetch_from_all_databases` functionality.
